mempool/src/shared_mempool/peer_manager.rs

```diff
@@ -124,6 +181,161 @@ impl PeerManager {
+            .get(peer)
+            .expect("missing peer info for peer")
+            .broadcast_info
+            .backoff_mode
+    }
+
+    /// Broadcasts txns to `peer`
+    ///
+    /// Broadcast strategy:
+    /// Prioritize resending older broadcasts that timed out (i.e. didn't receive ACK in set window of time)
+    /// or is retriable over sending more new broadcasts
+    /// 1. clear pending (i.e. unACK'ed) broadcasts if the corresponding timeline index read is empty
+    /// 2. find earliest broadcast that timed out or is retriable - rebroadcast if any
+    /// 3. else, broadcast fresh batch of txns
+    pub fn execute_broadcast<V>(
+        &self,
+        peer: PeerNetworkId,
+        scheduled_backoff: bool,
+        smp: &mut SharedMempool<V>,
+    ) where
+        V: TransactionValidation,
+    {
+        // start timer for tracking broadcast latency
+        let start_time = Instant::now();
+
+        let mut peer_info = self
+            .peer_info
+            .lock()
+            .expect("failed to acquire peer info lock");
+        let state = peer_info
+            .get_mut(&peer)
+            .expect("missing peer info for peer");
+
+        // only broadcast to peer that is both alive and picked
+        if !state.is_alive || !self.is_picked_peer(&peer) {
+            return;
+        }
+
+        // If backoff mode is on for this peer, only execute broadcasts that were scheduled as a backoff broadcast
+        // This is to ensure the backoff mode is actually honored (there is a chance a broadcast was scheduled
+        // in non-backoff mode before backoff mode was turned on - ignore such scheduled broadcasts)
+        if state.broadcast_info.backoff_mode && !scheduled_backoff {
+            return;
+        }
+
+        let mut mempool = smp
+            .mempool
+            .lock()
+            .expect("failed to acquire core mempool lock");
+
+        // Sync peer's pending broadcasts with latest mempool state
+        // A pending broadcast might become empty if the corresponding txns were committed through
+        // another peer, so don't track broadcasts for committed txns
+        state.broadcast_info.sent_batches = state
+            .broadcast_info
+            .sent_batches
+            .clone()
+            .into_iter()
+            .filter(|(id, _batch)| !mempool.timeline_range(id.0, id.1).is_empty())
+            .collect();
+
+        // check for batch to rebroadcast:
+        // 1. batch that did not receive ACK in configured window of time
+        // 2. batch that an earlier ACK marked as retriable
+        let rebroadcast = state
+            .broadcast_info
+            .sent_batches
+            .iter()
+            .find(|(_id, batch)| {
+                batch
+                    .send_time
+                    .checked_add(Duration::from_millis(
+                        self.mempool_config.shared_mempool_ack_timeout_ms,
+                    ))
```

Comment:
why are we doing check_add? if we overflow are we even in a good place any more? probably fine to either ignore the "potential" for an overflow or to just expect here...

Reply:
cc: @mimoo